buffling/snr.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Evaluate signal-to-noise ratio of the 'shuffling+flipping' algorithm."""
+
+import numpy as np
+import scipy as sp
+import pandas as pd
+
+# @title helpers
+def generate_transition_matrix(n_pub, p=0.1, first2=False, tol=1e-15):
+  '''Generate the transition matrix of logos. 
+
+  Args: 
+    n_pub: Integer, number of publishers.
+    p: Flipping probablity.
+    first2: Boolean, whether return only the first 2 rows.
+    tol: A small Real number. This will be add to the sum of each row, 
+      followed by a row normalizing. 
+  
+  Return: 
+    A numpy.array of n_pub x n_pub, containing the transition probability. 
+  '''
+  n_logo = 2**n_pub
+  n_row = 2 if first2 else n_logo
+  transition_mat = np.zeros((n_row, n_logo))
+  for i in range(n_row):
+    for j in range(n_logo):
+      n_p = bin(i ^ j)[2:].count('1')
+      transition_mat[i,j] = p**n_p * (1 - p)**(n_pub - n_p)
+  rs = transition_mat.sum(axis=1) + tol ## add a small number to row sums
+  return transition_mat / rs[:, np.newaxis]
+
+
+def calculate_snr_ij(i,j,input_logo_wgt,transition_mat,maximizer=False):
+  '''Calculate the SNR with fixed incremental change at i and j. 
```

Comment:
i and j in the description is not clear at first glance. May be something like "Calculate the SNR with fixed incremental change of two logo counts."?